benchmarks/src/main/scala/benchmarks/Benchmark.scala

```diff
@@ -0,0 +1,85 @@
+package benchmarks
+
+sealed abstract class BenchmarkResult(val name: String, val success: Boolean)
+
+case class BenchmarkCompleted(override val name: String,
+                              minNs: Long,
+                              maxNs: Long,
+                              avgNs: Long,
+                              iterations: Int,
+                              override val success: Boolean)
+    extends BenchmarkResult(name, success) {
+  override def toString: String = {
+    def format(n: Double, decimals: Int = 3): String = {
+      val s = n.toString
+      s.substring(0, s.indexOf('.') + decimals + 1)
+    }
+
+    val minMs = format(minNs / 1e6)
+    val maxMs = format(maxNs / 1e6)
+    val avgMs = format(avgNs / 1e6)
+    (if (success) "  [ok] " else "  [fail] ") + name +
+      s": $iterations iterations, min ${minMs}ms, max ${maxMs}ms," +
+      s" avg ${avgMs}ms"
+  }
+}
+
+case class BenchmarkFailed(override val name: String, cause: Throwable)
+    extends BenchmarkResult(name, false) {
+  override def toString: String =
+    s"""  [fail] $name threw an exception:
+       |$cause""".stripMargin
+}
+
+case class BenchmarkDisabled(override val name: String)
+    extends BenchmarkResult(name, true) {
+  override def toString: String =
+    s""" [???] $name is disabled."""
+}
+
+abstract class Benchmark[T] {
+  def run(): T
```

Comment:
As `run` does not take any arguments, wouldn't it be better to make in an entry point which is supposed to be run by `perf`? this will ensure that you don't need to invent your own measurement ways and ways to make sure you don't optimize acrross runs.